src/ImageSharp/Formats/Gif/LzwDecoder.cs

```diff
@@ -132,112 +132,140 @@ public void DecodePixelRow(Span<byte> indices)
                 {
-                    Unsafe.Add(ref pixelStackRef, (uint)this.top++) = Unsafe.Add(ref suffixRef, (uint)this.code);
-                    this.oldCode = this.code;
-                    this.first = this.code;
+                    pixelStack[top++] = suffix[code];
+                    oldCode = code;
+                    first = code;
                     continue;
                 }
 
-                int inCode = this.code;
-                if (this.code == this.availableCode)
+                int inCode = code;
+                if (code == availableCode)
                 {
-                    Unsafe.Add(ref pixelStackRef, (uint)this.top++) = (byte)this.first;
-
-                    this.code = this.oldCode;
+                    pixelStack[top++] = first;
+                    code = oldCode;
                 }
 
-                while (this.code > this.clearCode)
+                while (code > clearCode && top < maxTop)
                 {
-                    Unsafe.Add(ref pixelStackRef, (uint)this.top++) = Unsafe.Add(ref suffixRef, (uint)this.code);
-                    this.code = Unsafe.Add(ref prefixRef, (uint)this.code);
+                    pixelStack[top++] = suffix[code];
+                    code = prefix[code];
                 }
 
-                int suffixCode = Unsafe.Add(ref suffixRef, (uint)this.code);
-                this.first = suffixCode;
-                Unsafe.Add(ref pixelStackRef, (uint)this.top++) = suffixCode;
+                int suffixCode = suffix[code];
+                first = suffixCode;
+                pixelStack[top++] = suffixCode;
 
-                // Fix for Gifs that have "deferred clear code" as per here :
+                // Fix for GIFs that have "deferred clear code" as per:
                 // https://bugzilla.mozilla.org/show_bug.cgi?id=55918
-                if (this.availableCode < MaxStackSize)
+                if (availableCode < MaxStackSize)
                 {
-                    Unsafe.Add(ref prefixRef, (uint)this.availableCode) = this.oldCode;
-                    Unsafe.Add(ref suffixRef, (uint)this.availableCode) = this.first;
-                    this.availableCode++;
-                    if (this.availableCode == this.codeMask + 1 && this.availableCode < MaxStackSize)
+                    prefix[availableCode] = oldCode;
+                    suffix[availableCode] = first;
+                    availableCode++;
+                    if (availableCode == codeMask + 1 && availableCode < MaxStackSize)
                     {
-                        this.codeSize++;
-                        this.codeMask = (1 << this.codeSize) - 1;
+                        codeSize++;
+                        codeMask = (1 << codeSize) - 1;
                     }
                 }
 
-                this.oldCode = inCode;
+                oldCode = inCode;
             }
 
             // Pop a pixel off the pixel stack.
-            this.top--;
+            top--;
```

Comment:
Is there a some logical guarantee that `top` doesn't run below zero?

Reply:
As soon as we hit zero we enter the condition above. In there I believe there's at least one condition that increments `top` by one. The worst thing that could ever happen if `top` fals below zero is that we skip to the end of the pixel indices.